mesonbuild/interpreter/compiler.py

```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
         # other targets, list outputs, etc.
         private_dir = os.path.relpath(self.interpreter.backend.get_target_private_dir(tg), self.interpreter.subdir)
         return [build.CustomTargetIndex(tg, os.path.join(private_dir, o)) for o in tg.outputs]
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+                  KwargInfo('api', (str, NoneType)),
+                  KwargInfo('compilation', (str, NoneType)),
+                  KwargInfo('static_compilation', (str, NoneType)),
+                  KwargInfo('static_only', bool, default=False))
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '1.1.0')
+    def symbol_visibility_header_method(self, args: T.Tuple[str], kwargs: 'SymbolVisibilityHeaderKW') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp', 'objc', 'objcpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        if default_prefix[0].isdigit():
+            default_prefix = f'_{default_prefix}'
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = '#pragma once\n'
```

Comment:
I would rather not use `#pragma once` here. I think we should use include guards instead since they are much more universal.

Reply:
I never understood why people use header guards, as far as I know pragma once is universal and faster because it does not have to preprocess the whole file. It reads only the first line the first time it's included and next time it does not even open the file at all.

That said, I'm not against changing that here, should not make a real difference for a small file like that.

Reply:
No sense in changing it until there is more consensus on this PR

Reply:
Since we already produce pragma once for configure_file, it makes sense to produce it for anything else too as there is no advantage to doing so *inconsistently*.

Reply:
I didn't know there is precedent for using `#pragma once` in code generated by Meson. Since nobody ever complained, we can assume that's fine.

Reply:
I'm not certain whether anyone complained on the meson bugtracker.

Certainly, people complain on general C/C++ discussion sites that it isn't "portable" despite their refusal to use any compiler other than GCC, clang, or MSVC on account of the lot of the remainder being "half-baked garbage that doesn't implement standards from the last decade" and in some cases "is working towards c99 compliance but we're not quite there yet, please send patches".

Of course all 3 of those support pragma once.

Meanwhile, the standards committee doesn't want to standardize pragma once because they are concerned that the language specification shouldn't mandate that source code come inside files, because filesystems are mysterious beasts that are considered way too complicated for Serious Business™ like a programming language standard to depend on. What should the programming language describe as the correct way to handle headers that are a symlink? Or even worse, what about hardlinks??? No, no, this "filesystem" technology is too unpredictable.

Reply:
It's impossible to write C without hitting a million UB. At some point when all serious implementation does the same thing, who cares what the spec says...

Reply:
I find it hard to out into words, how utterly upset I was to (recently) discover that the spec says if you don't include an end of line character at the end of your file (the posix definition of a text, rather than binary file, is that it has posix line terminators)...

... Then because C/C++ source code is *text files*, and a file that isn't a posix text file is technically not C/C++ source code, the spec declares this, too, is UB. And not something sane like I don't know, an input processing "error: an input file was specified that isn't a source code file". Or even just saying "an implementation is permitted to accept inputs other than source files as an implementation extension, as long as it successfully parses as valid C/C++".

But no, they outright declare that compilers have a full license to generate UB and make demons fly out of your nose instead of doing what your code's C/C++ mapping says.

Nasal demons for your line endings. It's obscene. But they don't want to standardize pragma once ***because filesystems***.